### PR TITLE
[Part 4][Unit tests] JAMES-2717 Remove EmbeddedElasticSearch in unit tests

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java [hunk 1]
```diff
@@ -45,37 +41,39 @@
...
     private static final WriteAliasName ALIAS_NAME = new WriteAliasName("alias_name");
     private static final TypeName TYPE_NAME = new TypeName("type_name");
-    private TemporaryFolder temporaryFolder = new TemporaryFolder();
-    private EmbeddedElasticSearch embeddedElasticSearch = new EmbeddedElasticSearch(temporaryFolder);
 
-    @Rule
-    public RuleChain ruleChain = RuleChain.outerRule(temporaryFolder).around(embeddedElasticSearch);
-
-    private Node node;
+    private final DockerElasticSearch elasticSearch = DockerElasticSearchSingleton.INSTANCE;
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java]: Why don't we use an extension/rule for this?

File: mailbox/plugin/quota-search-elasticsearch/src/main/java/org/apache/james/quota/search/elasticsearch/events/ElasticSearchQuotaMailboxListener.java [hunk 2]
```diff
@@ -60,7 +60,9 @@ public boolean isHandling(Event event) {
 
     @Override
     public void event(Event event) throws JsonProcessingException {
-        handleEvent(event.getUser(), (QuotaUsageUpdatedEvent) event);
+        if (event instanceof QuotaUsageUpdatedEvent) {
```
Comment (chibenwa) [inline mailbox/plugin/quota-search-elasticsearch/src/main/java/org/apache/james/quota/search/elasticsearch/events/ElasticSearchQuotaMailboxListener.java]: How is this change related?

File: mailbox/plugin/quota-search-elasticsearch/src/test/java/org/apache/james/quota/search/elasticsearch/events/ElasticSearchQuotaMailboxListenerTest.java [hunk 3]
```diff
@@ -45,29 +45,25 @@
...
 
     private static final int BATCH_SIZE = 1;
     private static final Event DUMB_EVENT = mock(Event.class);
 
-    private TemporaryFolder temporaryFolder = new TemporaryFolder();
-    private EmbeddedElasticSearch embeddedElasticSearch = new EmbeddedElasticSearch(temporaryFolder);
-
-    @Rule
-    public RuleChain ruleChain = RuleChain.outerRule(temporaryFolder).around(embeddedElasticSearch);
+    private final DockerElasticSearch elasticSearch = DockerElasticSearchSingleton.INSTANCE;
```
Comment (chibenwa) [inline mailbox/plugin/quota-search-elasticsearch/src/test/java/org/apache/james/quota/search/elasticsearch/events/ElasticSearchQuotaMailboxListenerTest.java]: Use a rule

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/IndexCreationFactoryTest.java [hunk 4]
```diff
@@ -21,34 +21,31 @@
...
     public static final IndexName INDEX_NAME = new IndexName("index");
     public static final ReadAliasName ALIAS_NAME = new ReadAliasName("alias");
 
-    private TemporaryFolder temporaryFolder = new TemporaryFolder();
-    private EmbeddedElasticSearch embeddedElasticSearch = new EmbeddedElasticSearch(temporaryFolder);
-
-    @Rule
-    public RuleChain ruleChain = RuleChain.outerRule(temporaryFolder).around(embeddedElasticSearch);
-
+    private final DockerElasticSearch elasticSearch = DockerElasticSearchSingleton.INSTANCE;
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/IndexCreationFactoryTest.java]: Use a rule

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/NodeMappingFactoryTest.java [hunk 5]
```diff
@@ -21,31 +21,22 @@
...
     public static final ReadAliasName ALIAS_NAME = new ReadAliasName("alias");
     public static final TypeName TYPE_NAME = new TypeName("type");
 
-    private TemporaryFolder temporaryFolder = new TemporaryFolder();
-    private EmbeddedElasticSearch embeddedElasticSearch = new EmbeddedElasticSearch(temporaryFolder);
-
-    @Rule
-    public RuleChain ruleChain = RuleChain.outerRule(temporaryFolder).around(embeddedElasticSearch);
-
+    private final DockerElasticSearch elasticSearch = DockerElasticSearchSingleton.INSTANCE;
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/NodeMappingFactoryTest.java]: Use a rule

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/search/ScrollIterableTest.java [hunk 6]
```diff
@@ -57,25 +55,25 @@
     public static final ReadAliasName ALIAS_NAME = new ReadAliasName("alias");
     public static final TypeName TYPE_NAME = new TypeName("messages");
 
-    private TemporaryFolder temporaryFolder = new TemporaryFolder();
-    private EmbeddedElasticSearch embeddedElasticSearch = new EmbeddedElasticSearch(temporaryFolder);
-
-    @Rule
-    public RuleChain ruleChain = RuleChain.outerRule(temporaryFolder).around(embeddedElasticSearch);
-
+    private final DockerElasticSearch elasticSearch = DockerElasticSearchSingleton.INSTANCE;
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/search/ScrollIterableTest.java]: Use a rule

File: mailbox/plugin/quota-search-elasticsearch/src/test/java/org/apache/james/quota/search/elasticsearch/ElasticSearchQuotaSearchTestSystemExtension.java [hunk 7]
```diff
@@ -44,12 +44,10 @@
 import org.junit.jupiter.api.extension.ParameterContext;
 import org.junit.jupiter.api.extension.ParameterResolutionException;
 import org.junit.jupiter.api.extension.ParameterResolver;
-import org.junit.rules.TemporaryFolder;
 
 public class ElasticSearchQuotaSearchTestSystemExtension implements ParameterResolver, BeforeEachCallback, AfterEachCallback {
 
-    private TemporaryFolder temporaryFolder = new TemporaryFolder();
-    private EmbeddedElasticSearch embeddedElasticSearch = new EmbeddedElasticSearch(temporaryFolder);
+    private final DockerElasticSearch elasticSearch = DockerElasticSearchSingleton.INSTANCE;
```
Comment (chibenwa) [inline mailbox/plugin/quota-search-elasticsearch/src/test/java/org/apache/james/quota/search/elasticsearch/ElasticSearchQuotaSearchTestSystemExtension.java]: Use a rule

File: mpt/impl/imap-mailbox/elasticsearch/src/test/java/org/apache/james/mpt/imapmailbox/elasticsearch/host/ElasticSearchHostSystem.java [hunk 8]
```diff
@@ -65,30 +62,29 @@
...
     private StoreMailboxManager mailboxManager;
 
-
     @Override
     public void beforeTest() throws Exception {
         super.beforeTest();
-        this.tempDirectory = Files.createTempDirectory("elasticsearch");
-        this.embeddedElasticSearch = new EmbeddedElasticSearch(tempDirectory);
-        embeddedElasticSearch.before();
+        this.dockerElasticSearch = DockerElasticSearchSingleton.INSTANCE;
```
Comment (chibenwa) [inline mpt/impl/imap-mailbox/elasticsearch/src/test/java/org/apache/james/mpt/imapmailbox/elasticsearch/host/ElasticSearchHostSystem.java]: Use a rule

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTest.java [hunk 9]
```diff
@@ -66,18 +64,15 @@
...
-    private TemporaryFolder temporaryFolder = new TemporaryFolder();
-    private EmbeddedElasticSearch embeddedElasticSearch = new EmbeddedElasticSearch(temporaryFolder);
-
-    @Rule
-    public RuleChain ruleChain = RuleChain.outerRule(temporaryFolder).around(embeddedElasticSearch);
-
     @ClassRule
     public static TikaContainerSingletonRule tika = TikaContainerSingletonRule.rule;
+
+    private final DockerElasticSearch elasticSearch = DockerElasticSearchSingleton.INSTANCE;
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/ElasticSearchIntegrationTest.java]: Use a rule

File: mailbox/plugin/quota-search-elasticsearch/src/test/java/org/apache/james/quota/search/elasticsearch/events/ElasticSearchQuotaMailboxListenerTest.java [hunk 10]
```diff
@@ -47,27 +47,25 @@
...
 
     private static final int BATCH_SIZE = 1;
     private static final Event DUMB_EVENT = mock(Event.class);
 
-    private TemporaryFolder temporaryFolder = new TemporaryFolder();
-    private EmbeddedElasticSearch embeddedElasticSearch = new EmbeddedElasticSearch(temporaryFolder);
-
     @Rule
-    public RuleChain ruleChain = RuleChain.outerRule(temporaryFolder).around(embeddedElasticSearch);
+    public DockerElasticSearch elasticSearch = DockerElasticSearchSingleton.INSTANCE;
```
Comment (chibenwa) [inline mailbox/plugin/quota-search-elasticsearch/src/test/java/org/apache/james/quota/search/elasticsearch/events/ElasticSearchQuotaMailboxListenerTest.java]: Don't we have a rule leveraging the singleton instance?

Comment (trantienduchn) [reply mailbox/plugin/quota-search-elasticsearch/src/test/java/org/apache/james/quota/search/elasticsearch/events/ElasticSearchQuotaMailboxListenerTest.java]: there is `DockerElasticSearchRule` but it's more about guice + TestRule, and it's located in intergration test module

Comment (chibenwa) [reply mailbox/plugin/quota-search-elasticsearch/src/test/java/org/apache/james/quota/search/elasticsearch/events/ElasticSearchQuotaMailboxListenerTest.java]: Let's create another one just about ES and in backend ^^

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/search/ScrollIterableTest.java [hunk 11]
```diff
@@ -109,14 +109,16 @@ public void scrollIterableShouldWorkWhenOneElement() {
...
 
             SearchRequestBuilder searchRequestBuilder = client.prepareSearch(INDEX_NAME.getValue())
                 .setTypes(TYPE_NAME.getValue())
                 .setScroll(TIMEOUT)
                 .setQuery(matchAllQuery())
                 .setSize(SIZE);
-            assertThat(convertToIdList(new ScrollIterable(client, searchRequestBuilder))).containsOnly(id);
+
+            WAIT_CONDITION
+                .untilAsserted(() -> assertThat(convertToIdList(new ScrollIterable(client, searchRequestBuilder))).containsOnly(id));
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/search/ScrollIterableTest.java]: indent

Comment (chibenwa) [reply backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/search/ScrollIterableTest.java]: I would separete the assertion from the await...

**Comment by trantienduchn** [comment]
```
[b7b5e70bdd24f2156972e65c0428ff0fce634eaf] [ERROR]   FixingGhostMailboxTest.ghostMailboxBugShouldDiscardOldContent:205 1 expectation failed.
[b7b5e70bdd24f2156972e65c0428ff0fce634eaf] JSON path [0][1].messageIds doesn't match.
[b7b5e70bdd24f2156972e65c0428ff0fce634eaf] Expected: a collection with size <1>
[b7b5e70bdd24f2156972e65c0428ff0fce634eaf]   Actual: []
```
test this please

**Comment by trantienduchn** [comment]
rebased to avoid conflict

**Comment by trantienduchn** [comment]
rebase to pick the commit from #2349

File: mailbox/plugin/quota-search-elasticsearch/src/main/java/org/apache/james/quota/search/elasticsearch/events/ElasticSearchQuotaMailboxListener.java (hunk 2 above)
Comment (mbaechler) [inline mailbox/plugin/quota-search-elasticsearch/src/main/java/org/apache/james/quota/search/elasticsearch/events/ElasticSearchQuotaMailboxListener.java]: The change is wrong: the EventBus always call `isHandling()` before calling `event()`, if a test doesn't replicate this behavior, we need to change the test.

**Comment by trantienduchn** [comment]
:green_apple:

**Comment by chibenwa** [comment]
```
[576b29435a1d68900edb8d06b80865f5f481ce94] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-jar-plugin:3.0.2:test-jar (default) on project apache-mailet: Error assembling JAR: Failed to create directory: /james-project/mailet/target/maven-archiver -> [Help 1]
```

Can be considered :green_apple: (all tests OK and deployment validation issue here is a jenkins issue)

Test this please anyway

**Comment by trantienduchn** [comment]
test this please

**Comment by Arsnael** [comment]
```
[576b29435a1d68900edb8d06b80865f5f481ce94] CONFLICT (rename/rename): Rename "server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/TestEmbeddedElasticSearchModule.java"->"server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/mdc/RequestId.java" in branch "HEAD" rename "server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/TestEmbeddedElasticSearchModule.java"->"server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/CucumberElasticSearchSingleton.java" in "SHA1_BRANCH"
[576b29435a1d68900edb8d06b80865f5f481ce94] CONFLICT (rename/rename): Rename "server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/TestEmbeddedESMetricReporterModule.java"->"server/container/util/src/main/java/org/apache/james/util/InputStreamUtils.java" in branch "HEAD" rename "server/container/guice/cassandra-guice/src/test/java/org/apache/james/modules/TestEmbeddedESMetricReporterModule.java"->"backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/DockerElasticSearchRule.java" in "SHA1_BRANCH"
```
fix the conflict please

**Comment by chibenwa** [comment]
Merged
